src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,386 @@
+                "the algorithm for the SecretKey return value must not be null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                "the algorithm for the SecretKey return value must not be "
+                + "empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(derivationSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived keying material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationSpec)
+        throws InvalidAlgorithmParameterException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        byte[] inputKeyMaterial;
+        byte[] salt;
+        byte[] pseudoRandomKey;
+        byte[] info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (derivationSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) derivationSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine both of the above Lists of key
+            // segments into one SecretKey object each, unless we were passed
+            // something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // perform extract
+            try {
+                return hkdfExtract(inputKeyMaterial, salt);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidAlgorithmParameterException(
+                    "an HKDF Extract could not be initialized with the given "
+                    + "key or salt material", ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
```

Comment:
There's a bit of inconsistency in using capitalization on abbreviations, examples are: "HKDF" vs "Mac"/"Hmac" vs "ikm". I'd suggest aligning with prior art and use that consistently.